### ISSUE TITLE
[Performance] synchronous XHR calls

### ISSUE THREAD

**Issue by hrj** [jvilk/BrowserFS#127]
When instantiating a new XHR file system with:
```
new BrowserFS.FileSystem.XmlHttpRequest('listings.json', url)
```
a synchronous XHR is issued. When the response time from the server is large, it causes a noticeable freeze of the browser tab. A warning is also shown in Chromium's dev console:
```
Synchronous XMLHttpRequest on the main thread is deprecated because of its detrimental effects to
the end user's experience. For more help, check http://xhr.spec.whatwg.org/.
```

Can this synchronous call be avoided? (To keep the same API, the implementation could probably use a `Promise` internally.)

**Comment by hrj** [comment]
> (To keep the same API, the implementation could probably use a Promise internally.)

Just noticed that the typescript target is set to `ES3`! Hence, no generators/yield. The sync XHR is probably not as simple to eliminate as I thought.

One way could be to async load the `listing.json` and use that for the async FS API implementations, and use a sync XHR for sync API implementations, if not already loaded.

That is,
* an async XHR for `listing.json` is started upon initialisation, and a promise is created.
* the async FS API defers execution with `promise.then()`
* the sync FS API checks if `listing.json` is already available, else issues a sync XHR request for it. And then continues its usual sync operation.

**Comment by jvilk** [comment]
I want to change this interface so there's an asynchronous `instantiate` function or something to that effect, similar to the other backends. It would be a breaking change, but it's worth it imo.

**Comment by jvilk** [comment]
Actually, we could also accept an optional callback in the constructor. If presented, then do it async. Else, do it sync.

**Comment by jvilk** [comment]
I've been debating, as well, if it would be better to accept a URL to a JSON object or a JSON object directly. Some users may want to just include the listing in their JavaScript.

**Comment by hrj** [comment]
> Actually, we could also accept an optional callback in the constructor. If presented, then do it async. Else, do it sync.

That sounds good to me. Much simpler solution than the one I proposed (which was trying to stay within current API limits).

> I've been debating, as well, if it would be better to accept a URL to a JSON object or, optionally, a JSON object directly. Some users may want to just include the listing in their JavaScript.

That would be great and would be even more simple than the above. Those who require async loading from URL can simply issue the XHR themselves and pass it to the FS constructor.

**Comment by jvilk** [comment]
Agreed. Then perhaps here's the task list:

- [ ] Change XHR constructor to take in a JSON object of listings and a URL prefix.
- [ ] (Optional) Factory method to create XHRFS from URL to listings.
- [ ] (Optional) Make the listings object more concise, which requires changing (or rewriting) the listings generator tool.

**Comment by hrj** [comment]
I am working on this now.

>  Change XHR constructor to take in a JSON object of listings and a URL prefix.

So will this be an API change that bumps up the major version number?

**Comment by jvilk** [comment]
@hrj Since it is a breaking change in a commonly-used API, it will. I'll release it as `1.0.0`. It's about time I started using major version numbers appropriately and follow strict semantic versioning!

**Comment by jvilk** [comment]
Also, thanks again for all of your help. I really appreciate your contributions!

**Comment by hrj** [comment]
@jvilk NP. Wouldn't have been possible without your quick and detailed responses! Thanks!

I have raised a PR for the first item, except that the constructor accepts both string and object. The benefit is that there is no major API breakage. If this is accepted, (2) is not required.

(3) is a little out of my reach right now, so I will pass.

**Comment by jvilk** [comment]
That's fair.

Maybe when I do the next major release, I'll make the listings format a bit more tidy.

**Comment by hrj** [comment]
I believe we have fixed everything in PR #129 
Hence, closing.
